submissions/LuckyDnepr/memory-pair-game/script.js

```diff
@@ -0,0 +1,141 @@
+});
+
+makeMenuAndListeners();
+
+function gameStart() {
+    foundPairs = 0;
+    renderGameField(numberOfCards);
+    generateCardsSet(numberOfCards, themesData[themeIndex].pictures.length - 1);
+    switchMenuAndGameField();
+}
+
+function renderGameField(numberOfCards) {
+    const gameField = doc.querySelector(".game_field");
+    gameField.classList.remove("field3x2");
+    gameField.classList.remove("field4x3");
+    switch (numberOfCards) {
+        case 6:
+            gameField.classList.add("field3x2");
+            break;
+        case 12:
+            gameField.classList.add("field4x3");
+            break;
+        default:
+            break;
+    }
+    gameField.innerHTML = new Array(numberOfCards)
+        .fill(0)
+        .reduce((innerhtml, curr, i) => {
+            return (innerhtml += `<div class="game_card" data-card_id="${i}">
+                <img src="./images/bg-hearts.png" alt="" class="card_face card_face_front">
+                <img src="#" alt="" class="card_face card_face_back">
+            </div>
+        `);
+        }, "");
+}
+
+function cardFlip(card) {
+    if (!card.classList.contains("flipped")) {
+        card.classList.add("flipped");
+        openedCards.push(card);
+        changeCardFace(card);
+        checkPair();
+    }
+}
+
+function checkPair() {
+    if (openedCards.length === 2) {
+        const firstCard = openedCards.at(-1),
+            secondCard = openedCards.at(-2);
+        if (
+            cardsSet[firstCard.dataset.card_id] !==
+            cardsSet[secondCard.dataset.card_id]
+        ) {
+            noMatchFound(firstCard, secondCard);
+        } else {
+            matchFound(firstCard, secondCard);
+        }
+    }
+    winCheck();
+}
+
+function noMatchFound(firstCard, secondCard) {
+    openedCards = [];
+    setTimeout(() => {
+        firstCard.classList.remove("flipped");
+        secondCard.classList.remove("flipped");
+        setTimeout(() => {
+            changeCardFace(firstCard);
+            changeCardFace(secondCard);
+        }, 500);
+    }, 1000);
+}
+
+function matchFound(firstCard, secondCard) {
+    openedCards = [];
+    firstCard.classList.add("pair");
+    secondCard.classList.add("pair");
+    foundPairs++;
+}
+
+function winCheck() {
+    if (foundPairs === cardsSet.length / 2) {
+        setTimeout(() => {
+            toggleWinWindow();
+        }, 1000);
+    }
+}
+
+function resetCardsStage() {
+    document.querySelectorAll(".game_card").forEach((card) => {
+        card.classList.remove("flipped");
+        card.classList.remove("pair");
+    });
+}
+
+function generateCardsSet(sizeOfSet, highestNumber) {
+    let mixedSetOfPairs = [],
+        setOfPairs = [];
+    for (let i = 0; i < sizeOfSet / 2; i++) {
+        const randomNumber = Math.floor(Math.random() * highestNumber);
+        if (!setOfPairs.includes(randomNumber)) {
+            setOfPairs.push(randomNumber, randomNumber);
+        } else {
+            i--;
+        }
+    }
```

Comment:
As for me you should take a copy of the array with images shuffle it and take first `sizeOfSet / 2` elements and then work with this copy. What do you think about that?

Reply:
#### Perhaps.
If you take the first **sizeOfSet/2** elements, and then work with them.
You have to somehow get an array of **sizeOfSet**, in which the indexes of two cards are randomly distributed throughout the **sizeOfSet**. Again we'll have to shuffle everything.
If we simply glue the two shuffled **sizeOfSet/2** element arrays together, we get **card[i] === card[sizeOfSet/2 + i]**, which is too simple gameplay-wise.
I spent a lot of time thinking about randomizing the set of cards, and came to the conclusion that there's no right way to do it. You can generate a jumbled array of **2n** cards in different ways.